### ISSUE TITLE
ipad/iphone

### ISSUE THREAD

**Issue by holtzermann17** [ether/pad#61]
Support for iPad (this could be huge).

**Comment by laughinghan** [comment]
AFAIK the iPhone/iPad don't support contentEditable/designMode, so this is impossible without rewriting most of ACE

**Comment by tekacs** [comment]
Perhaps a way to turn off ACE and use simpler symbols to manipulate the text could be implemented? That is, using a standard text control...

I can see all sorts of issues with this, such as mismatched delimiters if things such as bold/underline are used, leading to discrepancies between the two interfaces but if, for example, similarly to line numbers, the option affected all users' view of the pad (i.e. plain text view for all), this shouldn't be a problem. A compromise, perhaps, but it may well work...

I'll have a go, but I can't promise anything... :P

**Comment by JohnMcLear** [comment]
I'm not interested in supporting a closed platform

**Comment by tekacs** [comment]
Well how about Android, Meego, ELinks and so on? The idea suggested here, I believe, is to keep the editor in-browser, rather than making a native app.
For that matter, I've got a branch with changes that go some way towards implementing the idea I suggested above but I've been too busy to finish it off - I might see if I can push it up into a branch on my fork in case anyone else has a chance to finish it off...

**Comment by JohnMcLear** [comment]
I aim to support Android, techkid email me at john@mclear.co.uk and I will show you what I have so far.

**Comment by theoryshaw** [comment]
Hi, just curious if there's been any developments with this?...and how involved it is?  I'm the one that commissioned Egil Möller's help on developing the SketchSpace add-on: https://github.com/redhog/pad/commits/SketchSpace.net

We have a number of architects/engineers using http://ss.openingdesign.com/ right now, and they are clamoring for an Ipad or Android version.

**Comment by theoryshaw** [comment]
fyi, the SketchSpace plug-in is built using the dojox.gfx vector graphics API. http://dojotoolkit.org/reference-guide/dojox/gfx.html

Doing a quick search, I found the following post and code for accommodating a touchscreen interface: http://wesleylauka.com/musings/notes/dojoxgfx-drag-and-drop-touchscreen

**Comment by laughinghan** [comment]
my observation is unfortunately still true, Mobile WebKit, which powers the iPad and Android web browsers, doesn't support one of the key technologies used by EtherPad. The only way I can imagine this being possible is to take the approach I took in MathQuill, which is to use a hidden textarea to capture text entry and use DOM manipulation to insert visible text in response.

It might be possible to do this in a way insulated from most of how EtherPad works, which would mean it's not actually that big a change, but I don't know to what extent that is the case so I can't forsee how much work this would be. I'd guess a lot, though.

**Comment by theoryshaw** [comment]
Thanks laughinghan.  

Ran across these too (might or 'might not' be helpful) 
http://stackoverflow.com/questions/723592/why-doesnt-contenteditable-work-on-the-iphone
http://accessgarage.wordpress.com/2009/05/08/how-to-hack-your-app-to-make-contenteditable-work/

**Comment by JohnMcLear** [comment]
We are still pondering creating a standalone Java app for Android / Desktop users @ PrimaryT.  We have 0 desire for supporting Apple devices.

**Comment by teachmemrlee** [comment]
I really wish you guys would reconsider this as i would love to use this in my class room but we adopted Mac, so we are 1 to 1 with the iPad in our high school.  this would be great for teaching kids to write effectively.

Thanks

**Comment by JohnMcLear** [comment]
I really wish Apple would reconsider not shipping their device with a crippled browser or even better than consumers got smart and started purchasing devices that return a better ROI from a learning / $ perspective. (I'm biased and these are my own opinions, not the opinion of the foundation)

It is impossible for me (personally) to fix this problem without an apple tablet.   

I  do not want to focus my efforts on something that only affects the wealthiest of our users, I'd rather focus on getting etherpad running perfectly on the OLPC devices which is exactly what we achieved with Etherpad Lite.

There is a chance I may look at this so we can support Android based tablets but imho tablets are consuming devices and there is no guarantee that when I fix Android tablet support it will fix Apple tablet support.

We support any device that has a physical keyboard (including amazon Kindle).  

To summarize, there are loads of apple developers out there that could probably make this fix but they are all too busy making 2$ apps that simulate fart noises.   They are probably in a better position than me to look at this problem, I try to avoid these people so it's really tough for me to resolve this!

**Comment by teachmemrlee** [comment]
Wow, I apologize.  

Some times we as educators have no say in the decisions on the tools that are given to us.  We simply have to use them as best we can.  I was introduced to you application and saw it as a way to help my students, who are by no means at the top of the economic charts.

Most of the people making farting apps are there to make money, not help people.  I am sorry if I thought you were trying to help.

Again i apologize if i have offended you with my comment, i am not angry just frustrated.  I will attempt to find a way to make this work.

Mr. Lee

**Comment by JohnMcLear** [comment]
I investigated this some more tonight..  I have written about it before: http://mclear.co.uk/2010/03/28/etherpad-on-android-based-devices/ and I think that's why I am getting a bit frustrated with the tiresome requests but no action.

It looks like Apple are adding contentEditable support to webkit in iOS5 according to TinyMCE however this might not be the case.  If they add contentEditable support to webkit that means chrome gets it too which means Android devices should get the functionality natively.

Source that is alive as authors website is down:  http://webcache.googleusercontent.com/search?q=cache:http://www.symphonious.net/2011/06/10/contenteditable-in-mobile-webkit-update/

Original source: http://www.symphonious.net/2011/06/10/contenteditable-in-mobile-webkit-update/

The Etherpad foundation has put a call out a number of time to partner with iPhone developers but no developers have come up with the goods after being given the run around several time.  With that in mind I don't feel like I want to invest any more of the foundations time in supporting such a small % of the market.  

As a foundation we are doing our best to support the majority of our users, unfortunately for you that excludes mobile users and this is due to a design choice(when you sit down and think about it for a while it was probably the right choice) made by Apple(mostly).  

If you can find some examples of contentEditable working on any other websites on mobile devices I would be interested to see how they have implemented it without creating some sort of psuedo replica keyboard. 

I would absolutely love it if you can persuade an iPhone &/or Android developer to make an app that works with Etherpad, if you want to take that on as a personal challenge I would be delighted and would extend my support.

**Comment by JohnMcLear** [comment]
Android users without a physical keyboard can bring up the virtual keyboard with a long press of the “Menu” key.

It is still a bug that the keyboard doesn’t appear by itself. Android users may be unaware that this is possible. What’s more, when the keyboard is brought up manually the browser is unaware of it and thus won’t force the page into the correct position to be able to see the text you’re typing.

**Comment by laughinghan** [comment]
wait, but does contentEditable/designMode work on Android devices with a physical keyboard/manually brought up virtual keyboard?

**Comment by JohnMcLear** [comment]
Yes.  For example it works on an Amazon Kindle as it has a physical keyboard.  I know it seems crazy but that's the way the cookie crumbles.  Basically the mobile browsers disable interaction with contentEditable/designMode through a touch screen.  Again I want to emphasize this is by design, it's not a bug.

**Comment by JohnMcLear** [comment]
Probably worth noting I'm not expert in this area, it would be nice to get some input from some web browser developers.

**Comment by BriSeven** [comment]
Etherpad lite works fine on the iPad.

**Comment by JohnMcLear** [comment]
Etherpad lite works on iOS5 devices however please ensure if your etherpad lite is public facing you provide a way to get feedback from your users.

I use idea informer to gather feedback, it's free, quick and easy.
